### ISSUE TITLE
How forecast multivariate time series?

### ISSUE THREAD

**Issue by valer1435** [valer1435/RepoPilot#35]
What to do if I want to forecast multivariate time series?

**Comment by RepoPilotAssistant** [comment]
👋 Welcome to the Issue Discussion!

Hello there! I'm here to assist you with any questions. 

Please feel free to share your thoughts, and I'll do my best to provide you with the information or support you need. Let's work together to resolve any issues and make things better!

Looking forward to your input! 🌟. To ask me just tag @RepoPilotAssistant

**Comment by valer1435** [comment]
@RepoPilotAssistant please answer

**Comment by valer1435** [comment]
@RepoPilotAssistant please answer

**Comment by RepoPilotAssistant** [comment]
To forecast multivariate time series using FEDOT, you can use either the Multi Time-Series Approach or the Multi-Modal Approach. The Multi Time-Series Approach involves training a model on multiple time-series of the same variable, transforming them into table data using lagged transformation, and then applying regression models. The Multi-Modal Approach treats each time-series as a separate data source node and leverages the relationships between different time-series to improve forecasting accuracy. Both methods involve transforming the time-series data into a format suitable for regression models to generate forecasts.

**Please follow links below:**
1. [Plot of the forecast:](https://fedot.readthedocs.io/en/latest/basics/ts_forecasting.html)
2. [Plot of the forecast:](https://fedot.readthedocs.io/en/latest/basics/ts_forecasting.html)
3. [Time Series Forecasting — FEDOT 0.7.4 documentation](https://fedot.readthedocs.io/en/latest/basics/ts_forecasting.html)
4. [Time Series Forecasting](https://fedot.readthedocs.io/en/latest/basics/ts_forecasting.html)
5. [Additionally you can examine papers about Fedot performance on different time series forecasting tasks [1] , [2], [3],](https://fedot.readthedocs.io/en/latest/benchmarks/forecasting.html)

**Comment by valer1435** [comment]
Give an example of Multi Time-Series Approach @RepoPilotAssistant

**Comment by RepoPilotAssistant** [comment]
To provide an example of the Multi Time-Series Approach using FEDOT, you can follow these steps:

1. **Define the target column and task parameters**: Specify the column with the target time-series and set the forecast length.
2. **Load data**: Use `InputData.from_csv_multi_time_series` to load data from a CSV file, specifying the task, target column, and columns to use.
3. **Split data**: Split the data into training and testing sets using `train_test_data_setup`.
4. **Initialize the model**: Create an instance of `Fedot` for time-series forecasting, passing the problem type, task parameters, and other settings.
5. **Fit the model**: Run the AutoML model design to fit the model on the training data.
6. **Generate forecasts**: Use the fitted model to generate forecasts on the test data.
7. **Evaluate metrics**: Calculate and print metrics such as RMSE, MAE, and MAPE to evaluate the model's performance.

This approach leverages multiple time-series to improve the model's ability to forecast the target time-series.

**Please follow links below:**
1. [Plot of the forecast:](https://fedot.readthedocs.io/en/latest/basics/ts_forecasting.html)
2. [Time series forecasting example — FEDOT 0.7.4 documentation](https://fedot.readthedocs.io/en/latest/examples/ts_forecasting.html)
3. [Plot of the forecast:](https://fedot.readthedocs.io/en/latest/basics/ts_forecasting.html)
4. [Time Series Forecasting — FEDOT 0.7.4 documentation](https://fedot.readthedocs.io/en/latest/basics/ts_forecasting.html)
5. [Notebooks — FEDOT 0.7.4 documentation](https://fedot.readthedocs.io/en/latest/examples/notebooks.html)

**Comment by valer1435** [comment]
Can you provide code example @RepoPilotAssistant

**Comment by RepoPilotAssistant** [comment]
To provide a code example of the Multi Time-Series Approach using FEDOT, you can follow these steps:

1. **Prepare Input Data**: Ensure your input data is structured as a multivariate time series.
2. **Initialize the Transformation**: Create an instance of `LaggedTransformationImplementation`.
3. **Apply Transformation**: Use the `fit` method to apply the lagged transformation on the input data.
4. **Stack Transformed Features**: Use the `stack_by_type_fit` method to stack the transformed features for multi-time series data.

Here's a simplified example:

```python
from fedot.core.data.data import InputData
from fedot.core.operations.evaluation.operation_implementations.data_operations.ts_transformations import LaggedTransformationImplementation
import numpy as np

# Example input data
features = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])  # Example multivariate time series
target = np.array([10, 11, 12])  # Example target values
forecast_length = 2
old_idx = np.array([0, 1, 2])

# Create InputData object
input_data = InputData(features=features, target=target, idx=old_idx)

# Initialize the transformation
lagged_transform = LaggedTransformationImplementation(window_size=2)

# Apply the transformation
transformed_target, transformed_idx = lagged_transform.fit(input_data, features, target, forecast_length, old_idx)

# Stack the transformed features
stacked_features, stacked_target, stacked_idx = lagged_transform.stack_by_type_fit(
    input_data, transformed_target, transformed_idx, features, target, old_idx
)

# Now you can use stacked_features and stacked_target for further processing or modeling
```

In this example, the `LaggedTransformationImplementation` is used to transform the multivariate time series data, and the `stack_by_type_fit` method is used to stack the transformed features for multi-time series data. This approach is particularly useful for preparing data for time series forecasting models.

**Please follow links below:**
1. [ts_transformations.py](https://github.com/aimclub/fedot/blob/master/fedot/core/operations/evaluation/operation_implementations/data_operations/ts_transformations.py)
2. [time_series.py](https://github.com/aimclub/fedot/blob/master/fedot/core/operations/evaluation/time_series.py)
3. [time_series.py](https://github.com/aimclub/fedot/blob/master/fedot/core/operations/evaluation/time_series.py)
4. [main.py](https://github.com/aimclub/fedot/blob/master/fedot/api/main.py)
5. [naive.py](https://github.com/aimclub/fedot/blob/master/fedot/core/operations/evaluation/operation_implementations/models/ts_implementations/naive.py)

**Comment by valer1435** [comment]
This example is wrong, can you look in docs for relevant example? @RepoPilotAssistant

**Comment by RepoPilotAssistant** [comment]
To provide a correct example of the Multi Time-Series Approach using FEDOT, you can follow these steps:

1. **Define the target column and task parameters**:
   ```python
   target = 'col_3'
   task = Task(TaskTypesEnum.ts_forecasting, TsForecastingParams(forecast_length=20))
   ```

2. **Load data from a CSV file**:
   ```python
   data = InputData.from_csv_multi_time_series(
       file_path='time_series.csv',
       task=task,
       target_column=target,
       columns_to_use=['col_1', 'col_2', 'col_3', ..., 'col_n'])
   ```

3. **Split the data for training and testing**:
   ```python
   train_data, test_data = train_test_data_setup(data, validation_blocks=2)
   ```

4. **Initialize the FEDOT model for time-series forecasting**:
   ```python
   model = Fedot(problem='ts_forecasting', task_params=task.task_params, timeout=5, n_jobs=-1, cv_folds=2)
   ```

5. **Run the AutoML model design and fit the model**:
   ```python
   pipeline = model.fit(train_data)
   pipeline.show()
   ```

6. **Obtain the forecast and evaluate the model**:
   ```python
   forecast = model.predict(test_data, validation_blocks=2)
   print(model.get_metrics(metric_names=['rmse', 'mae', 'mape'], target=test_data.target, validation_blocks=2))
   ```

This example demonstrates how to use multiple time-series to improve the forecast accuracy by leveraging the expanded training sample.

**Please follow links below:**
1. [Plot of the forecast:](https://fedot.readthedocs.io/en/latest/basics/ts_forecasting.html)
2. [Time series forecasting example — FEDOT 0.7.4 documentation](https://fedot.readthedocs.io/en/latest/examples/ts_forecasting.html)
3. [Plot of the forecast:](https://fedot.readthedocs.io/en/latest/basics/ts_forecasting.html)
4. [Time Series Forecasting — FEDOT 0.7.4 documentation](https://fedot.readthedocs.io/en/latest/basics/ts_forecasting.html)
5. [Notebooks — FEDOT 0.7.4 documentation](https://fedot.readthedocs.io/en/latest/examples/notebooks.html)
